File: embedded/python/cense_dec.py
import sys
import numpy as np
from librosa import core
import matplotlib.pyplot as plt
import soundfile as sf

# Constants: process parameters
sr = 32000
l_frame = 4096
l_hop = l_frame
q = 8

# Load code
d_filename = sys.argv[1]
c_file = open("".join((d_filename,"_enc.txt")), 'r')

# Load approximate third-octave band to spectrogram weights
iH = np.genfromtxt("tob_4096_i.txt", delimiter=',')

# Load Huffman dictionary
d_sym = []
d_code = []
with open("dict.txt") as d_file:
	for line in d_file: # = For each symbol
		line = line.strip()
		d_temp = line.split(',')
		# First expression is the symbol
		d_sym.append(int(d_temp[0]))
		# The rest is the code in binary
		d_code.append("".join(d_temp[1:]))

# Declarations/Initialisations
w = np.ones(l_frame)
l_tf = 4
ind_f = 0
q_norm = np.zeros(2)
ind_tf = 0
fft_norm = np.sum(np.square(w))/l_frame
X_st = np.empty((int(l_frame/2+1),0))

for line in c_file: # Texture frames
	# Read file
	line = line.strip()
	c_temp = line.split(',')
	q_norm[0] = float(c_temp[0])
	q_norm[1] = float(c_temp[1])
	X_huff_l = int(c_temp[2])
	X_huff = c_temp[3]
	# Huffman
	X_delta = []

	while X_huff:
		found = 0
		for ind_sym_dict in range(0,len(d_sym)):
			if X_huff.startswith(d_code[ind_sym_dict]):
				X_delta = np.append(X_delta,d_sym[ind_sym_dict])
				X_huff = X_huff[len(d_code[ind_sym_dict]):]
				found=1
		if (found == 0):
		#didn't found the symbol check if this is the end
			if (len(X_huff) <=  7):
				#ignore the last symbols in this case
				X_huff = [] 
					
	X_delta = [int(i) for i in X_delta]
	X_delta = np.reshape(X_delta,(-1,iH.shape[1])).transpose()
	print(X_delta)
	# Delta
	X_tob = np.zeros(X_delta.shape)
	X_tob[:,0] = X_delta[:,0]
	for ind_f_tf in range(1,X_delta.shape[1]):
		X_tob[:,ind_f_tf] = X_delta[:,ind_f_tf]+X_tob[:,ind_f_tf-1]
	# Normalisation + Quantization
	X_tob = q_norm[1]*X_tob/(np.power(2,q-1)-1)+q_norm[0]
	
	print(X_tob)

	# Third-octave bands to spectrogram
	X_tob = np.power(10,X_tob/10) # Linear amplitude
	X = np.dot(iH,X_tob)
	# Magnitude spectrogram to STFT
	X = np.sqrt(X*fft_norm)
	if X_st.ndim == 1:
		X_st = X
	else:
		X_st = np.concatenate((X_st, X), axis=1)
	
c_file.close()


# Phase recovery and signal synthesis with G&L - Needs 50% overlap minimum
n_iter = 20
n_frames = X_st.shape[1]
x_temp = np.random.randn(l_frame+(n_frames-1)*l_hop)
print(x_temp.shape)
for gl_iter in range(0,n_iter):
	x_stft = core.stft(x_temp, n_fft=l_frame, hop_length=l_hop+1, win_length=None, window='boxcar', center=True, dtype=None, pad_mode='reflect')
	b = X_st*x_stft/(np.absolute(x_stft)+1e-15)
	x_temp = core.istft(b, hop_length=l_hop+1, win_length=l_frame, window='boxcar', center=True, dtype=None, length=None)
		

x = x_temp/np.absolute(x_temp).max()

# Save resulting audio
np.savetxt("".join((d_filename,"_dec.txt")), x, fmt='%.18e', delimiter=',', newline='\n', header='', footer='', comments='# ')


sf.write('sinus_440_emb_dec.ogg', x, sr)


t = np.linspace(0,x_temp.shape[0]/sr, num=x_temp.shape[0])
plt.plot(t, x_temp)
plt.show()

File: embedded/matlab/cense_dec.m
% Needed for decoding: 
% - Encoded data (X_huff)
% - Normalisation values (q_norm)
% - Length of the encoded data (X_huff_l)

% ----- Function ----- %
% function cense_dec(X_huff, X_Huff_l, q_norm)
% ----- END - Function ----- %
% ----- Script ----- %
clear all, close all, clc;
load data_enc;
% ----- END - Script ----- %

sr = 32000; % Sampling rate (Hz)
l_frame = 4096; % STFT frame duration, sr/8 rounded to next power of 2 - approx. 128 ms
l_hop = l_frame; % STFT hop size
q = 8; % Quantized word size

% ----- Matlab ----- %
% The following steps are implemented in matlab only

%% Weight matrix, constant
N = 2^(ceil(log2(l_frame)));
N_filt = 2^20; % Design with a much higher precision
G = third_octave_filterbank(sr, N_filt, -17:11, 3);
G = G.^2;
G_mean(:, 1) = mean(G(:, 1:1+round(N_filt/(2*N))), 2);
G_mean(:, N/2+1) = mean(G(:, 1+(N/2)*round(N_filt/N)-round(N_filt/(2*N)):end), 2);
for ind_f = 2:N/2
    G_mean(:, ind_f) = mean(G(:, 1+(ind_f-1)*round(N_filt/N)-round(N_filt/(2*N)):1+(ind_f-1)*round(N_filt/N)+round(N_filt/(2*N))), 2);
end
H_st = G_mean;
clear ind_f N_filt G G_mean;

% Approximate inverse operation
HH = H_st'*H_st;
iH = H_st'./(repmat(max(mean(diag(HH))/100, sum(HH))',1,size(H_st, 1))); % Inverse frequency weights matrix
iH(isnan(iH)) = 0; % Some frequencies are outside analysis range, ie we divide by zero

%% Huffman dictionnary
load dict;

imp_ver = 1;
% ----- END - Matlab ----- %

if ~imp_ver
    %% All-at-once version
    
    % w = hanning(l_frame, 'periodic');
    w = rectwin(l_frame);
    fft_norm = sum(w.^2)*(l_frame/2+1)/(l_frame^2);

    %% Huffman decoding
    X_huff = X_huff(1:X_huff_l);
    X_delta = huffmandeco(X_huff, dict);
    X_delta = reshape(X_delta, size(H_st, 1), []); % To matrix
    
    %% Delta decoding
    X_tob = zeros(size(X_delta));
    prev = zeros(size(X_delta, 1), 1);
    for ind_frame = 1:size(X_delta, 2);
        X_tob(:, ind_frame) = X_delta(:, ind_frame) + prev;
        prev = X_tob(:, ind_frame);
    end
    
    %% Quantization
    X_tob = q_norm(2)*X_tob./(2^(q-1)-1)+q_norm(1);
    
    %% Third-octave bands to spectrogram
    X_tob = 10.^(X_tob/10);
    X_st = iH*X_tob; % Estimate the spectrogram
    
    %% Magnitude spectrogram to STFT
    
    X_st = X_st*fft_norm; % Normalize to conserve the energy of x
    X_st = sqrt(X_st);
    
    n_iter = 20;
    n_frames = size(X_st, 2);
    x_temp = randn(l_frame+(n_frames-1)*l_hop, 1); % Random values initialisation
    iter = 1;
    while iter < n_iter+1 % Griffin and Lim
        disp(['Iteration ' num2str(iter) ' of ' num2str(n_iter) '...']);
        iter = iter+1;
        x_stft = stft(x_temp, l_frame, l_hop, 'rect', 0);
        b = sqrt(sum(w(1:end/2+1)))*X_st.*x_stft./(abs(x_stft)+eps); % As described in G&L paper
        % b = (sqrt(x_spec)/32768).*exp(1i*angle(x_stft)); % Also works
        x_temp = istft(b, l_frame, l_hop, 'rect');
    end
    x_r = x_temp./max(abs(x_temp));
    
else
    %% Alternative version - Texture frames
    % A certainly more realistic implementation would process each frame
    % sequentially, then group windows in texture frames for encoding
    % and transmission

    w = rectwin(l_frame);
    fft_norm = sum(w.^2)*(l_frame/2+1)/(l_frame^2);

    
    
    X_st_all = [];
    l_tf = 8; % Texture frame length in windows
    for ind_tf = 1:length(X_huff_l)
        %% Huffman decoding
        X_delta = huffmandeco(X_huff{ind_tf}(1:X_huff_l{ind_tf}), dict);
        X_delta = reshape(X_delta, size(H_st, 1), []); % To matrix
        
        %% Delta decoding
        X_tob = zeros(size(X_delta));
        prev = zeros(size(X_delta, 1), 1);
        for ind_frame = 1:size(X_delta, 2);
            X_tob(:, ind_frame) = X_delta(:, ind_frame) + prev;
            prev = X_tob(:, ind_frame);
        end
        
        %% Quantization
        X_tob = q_norm{ind_tf}(2)*X_tob./(2^(q-1)-1)+q_norm{ind_tf}(1);
        
        %% Third-octave bands to spectrogram
        X_tob = 10.^(X_tob/10);
        X_st = iH*X_tob; % Estimate the spectrogram
        
        %% Magnitude spectrogram to STFT
        X_st = X_st*fft_norm; % Normalize to conserve the energy of x
        X_st_all(:, end+1:end+size(X_st, 2)) = sqrt(X_st);
    end
    
    n_iter = 20;
    n_frames = size(X_st_all, 2);
    x_temp = randn(l_frame+(n_frames-1)*l_hop, 1); % Random values initialisation
    iter = 1;
    while iter < n_iter+1 % Griffin and Lim
        disp(['Iteration ' num2str(iter) ' of ' num2str(n_iter) '...']);
        iter = iter+1;
        x_stft = stft(x_temp, l_frame, l_hop, 'rect', 0);
        b = sqrt(sum(w(1:end/2+1)))*X_st_all.*x_stft./(abs(x_stft)+eps); % As described in G&L paper
        % b = (sqrt(x_spec)/32768).*exp(1i*angle(x_stft)); % Also works
        x_temp = istft(b, l_frame, l_hop, 'rect');
    end
    x_r = x_temp./max(abs(x_temp));
    
end
